### ISSUE TITLE
Equivalent to react controlled-component

### ISSUE THREAD

**Issue by vthinkxie** [angular/angular#42126]
# Feature Request
In React, the state be the “single source of truth” in [controlled components](https://reactjs.org/docs/forms.html#controlled-components)

but there is no equivalent feature in angular

### Relevant Package
`@angular/form`

### Description

- in React

   DEMO: https://stackblitz.com/edit/react-ts-kyqxmj?file=index.tsx

   type something in the `input`, and the input value won't change, the state be the single source of truth

- in Angular

   DEMO: https://stackblitz.com/edit/angular-ivy-sazmxc?file=src%2Fapp%2Fapp.component.ts

   type something in the `input`, and the input will change, the state won't be the single source of truth

### Describe the solution you'd like

support changing the input and other form components to use controlled mode or not controlled mode

**Comment by petebacondarwin** [comment]
This can be achieved already in a couple of ways. For instance https://stackblitz.com/edit/angular-ivy-wtkqqx?file=src%2Fapp%2Fapp.component.ts

I am not sure that there needs to be any additional change to how Angular input event handling works.

**Comment by alcfeoh** [comment]
Isn't it what `[(ngModel)]` does? And React has to "emulate" it through controlled components as it doesn't have two-way bindings?

**Comment by vthinkxie** [comment]
> Isn't it what `[(ngModel)]` does? And React has to "emulate" it through controlled components as it doesn't have two-way bindings?

actually no, imagine there is a checkbox and when you click on it, the final state needs to be determined by the backend, rather than being checked immediately after clicking.

this is what controlled components can do and ngModel can not.

**Comment by alcfeoh** [comment]
It can. You can use a Typescript getter / setter to register side effects on read / write of any class property.

**Comment by vthinkxie** [comment]
> It can. You can use a Typescript getter / setter to register side effects on read / write of any class property.

the side effects only work on the second render frame, which means the check box will flicker from 

in ngModel

click -> true -> response -> false

in controlled-components

click -> response -> false

**Comment by vthinkxie** [comment]
> This can be achieved already in a couple of ways. For instance https://stackblitz.com/edit/angular-ivy-wtkqqx?file=src%2Fapp%2Fapp.component.ts
> 
> I am not sure that there needs to be any additional change to how Angular input event handling works.

@petebacondarwin what about adding an option to ngModelOptions to change the behavior

**Comment by petebacondarwin** [comment]
I don't think we need to make any changes to the core framework. You can achieve the desired effect just using the standard Angular building blocks. Here is another example using a directive: https://stackblitz.com/edit/angular-ivy-vtshbt?file=src%2Fapp%2Fapp.component.ts

**Comment by petebacondarwin** [comment]
That being said, I think this sort of thing is a bad UX. As a user, if I interact with an input, then I expect it to follow what I intended, and if the server things something different, then I would set some indeterminate state on the input during that period. Perhaps this would be to disable the input while we do the lookup. This informs the user that although they have tried to change something it might not work and also that they have to wait for any update. I might even consider a spinner type device to reinforce that.

**Comment by waterplea** [comment]
> @petebacondarwin
> 
> thanks, but I think it is a little hack and needs extra codes to make it work

How do you think React does it? :) It does the same thing, but under the hood. Which is also what anybody can do if they make a custom directive for that. This is not something a framework has to care about because all the needed tools are already in place.

As for server response and user interaction. In Taiga UI we created a toggle with loader for this purpose:
https://stackblitz.com/edit/taiga-toggle-loader

I've made a very basic imitation of server request and sometimes failing to set the value to true, click it a few times.

**Comment by vthinkxie** [comment]
> > @petebacondarwin
> > thanks, but I think it is a little hack and needs extra codes to make it work
> 
> How do you think React does it? :) It does the same thing, but under the hood. Which is also what anybody can do if they make a custom directive for that. This is not something a framework has to care about because all the needed tools are already in place.
> 
> As for server response and user interaction. In Taiga UI we created a toggle with loader for this purpose:
> https://stackblitz.com/edit/taiga-toggle-loader
> 
> I've made a very basic imitation of server request and sometimes failing to set the value to true, click it a few times.

We have implemented the same component, but the toggle should not change until the server response

https://stackblitz.com/edit/angular-3gwqek?file=src/app/app.component.ts

What I want to proposal is that there should be an option to define the behavior, in react there is a [defaultValue](https://reactjs.org/docs/uncontrolled-components.html#default-values)

**Comment by vthinkxie** [comment]
> I don't think we need to make any changes to the core framework. You can achieve the desired effect just using the standard Angular building blocks. Here is another example using a directive: https://stackblitz.com/edit/angular-ivy-vtshbt?file=src%2Fapp%2Fapp.component.ts

Thanks @petebacondarwin 

I think it would be much better if the angular provide an official option to define the behavior instead of letting user customize it.

The defaultValue Document in React: https://reactjs.org/docs/uncontrolled-components.html#default-values

**Comment by slamjs** [comment]
what purpose does `[value]` serves in angular then? To provide an initial value? It's confusing and ridiculous.

`[value]` should provide a contolled component functionality, obviously.

**Comment by waterplea** [comment]
`[value]` is a native property `value` binding and that's what it does. It's not in Angular, it's in the DOM API :) You guys are literally asking for a one-liner to be added to the library. To simplify @petebacondarwin example, here's the shortest form of this directive:
```ts
@Directive({
  selector: 'input[controlled][type="checkbox]',
  host: {
    '(input)': '$event.target.checked = !$event.target.checked'
  }
})
export class ControlledCheckbox {}
```
Besides being questionable UX, this would probably get you in some trouble from accessibility point of view. This is altering native browser behavior, or rather reverting it which is probably even worse because of Ctrl+Z if you would try to use it on something, other than checkboxes.

I would argue this doesn't belong in the framework.
